File: 2022/3.hs
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
import Framework
import Data.List
import Data.Maybe ( fromJust )

only [x] = x

priority :: String -> Int
priority = fromJust . flip elemIndex (['a' .. 'z'] ++ ['A' .. 'Z']) . only . nub

splitIntoHalf :: [a] -> ([a], [a])
splitIntoHalf line = splitAt (length line `div` 2) line

day1 :: String -> Int
day1 = sum . map (priority . uncurry intersect . splitIntoHalf) . lines

findBadges :: [String] -> [Int]
findBadges [] = []
findBadges linesLeft =
  let (group, rest) = splitAt 3 linesLeft
      badge = foldr1 intersect group in
  priority badge : findBadges rest

day2 = sum . findBadges . lines

File: 2020/3.py
from framework import *

s = read_input_lines()

N = len(s[0])

x = 0
y = 0
t = 0
while y < len(s):
    x += 3
    y += 1
    if y < len(s): t += s[y][x % N] == '#'
print(t)


m = 1
for dx, dy in [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]:
    x = 0
    y = 0
    t = 0
    while y < len(s):
        x += dx
        y += dy
        if y < len(s): t += s[y][x % N] == '#'
    m *= t
print(m)
